core/arch/aarch64/proc.c

```diff
@@ -293,71 +265,97 @@ enable_all_test_cpu_features()
         FEATURE_BF16,   FEATURE_I8MM,       FEATURE_F64MM,   FEATURE_FlagM,
         FEATURE_JSCVT,  FEATURE_DPB,        FEATURE_DPB2,    FEATURE_SVE2,
         FEATURE_SVEAES, FEATURE_SVEBitPerm, FEATURE_SVESHA3, FEATURE_SVESM4,
-        FEATURE_MTE,    FEATURE_BTI,        FEATURE_FRINTTS
+        FEATURE_MTE,    FEATURE_BTI,        FEATURE_FRINTTS, FEATURE_PAUTH2
     };
     for (int i = 0; i < BUFFER_SIZE_ELEMENTS(features); ++i) {
         proc_set_feature(features[i], true);
     }
     dr_set_sve_vector_length(256);
 }
 
+uint64
+get_reg_val(feature_reg_idx_t feat_reg)
+{
+    return cpu_info.features.isa_features[feat_reg];
+}
+
 bool
-proc_has_feature(feature_bit_t f)
+proc_has_feature_imp(feature_bit_t feature_bit)
 {
-#ifndef DR_HOST_NOT_TARGET
-    ushort feat_nibble, feat_val, freg_nibble, feat_nsflag;
-    uint64 freg_val = 0;
-
-    feature_reg_idx_t feat_reg = GET_FEAT_REG(f);
-    switch (feat_reg) {
-    case AA64ISAR0: {
-        freg_val = cpu_info.features.flags_aa64isar0;
-        break;
-    }
-    case AA64ISAR1: {
-        freg_val = cpu_info.features.flags_aa64isar1;
-        break;
-    }
-    case AA64PFR0: {
-        freg_val = cpu_info.features.flags_aa64pfr0;
-        break;
-    }
-    case AA64MMFR1: {
-        freg_val = cpu_info.features.flags_aa64mmfr1;
-        break;
-    }
-    case AA64DFR0: {
-        freg_val = cpu_info.features.flags_aa64dfr0;
-        break;
-    }
-    case AA64ZFR0: {
-        freg_val = cpu_info.features.flags_aa64zfr0;
-        break;
-    }
-    case AA64PFR1: {
-        freg_val = cpu_info.features.flags_aa64pfr1;
-        break;
+    bool has_feature = false;
+
+    uint64 reg_val = get_reg_val(GET_FEAT_REG(feature_bit));
+    ushort nibble_pos = GET_FEAT_NIBPOS(feature_bit);
+    ushort reg_nibble = (reg_val >> (nibble_pos * 4)) & 0xFULL;
+
+    bool feat_nsflag = GET_FEAT_NSFLAG(feature_bit);
+    ushort feat_nibble = GET_FEAT_VAL(feature_bit);
+    bool exact_match = GET_FEAT_EXACT_MATCH(feature_bit);
+
+    if (feat_nsflag) {
+        // special case where a value of 0xF signifies the feature is not present
+        has_feature = (reg_nibble != 0xF);
+    } else {
+        if (exact_match)
+            has_feature = (reg_nibble == feat_nibble);
+        else
+            has_feature = (reg_nibble >= feat_nibble);
     }
-    default: CLIENT_ASSERT(false, "proc_has_feature: invalid feature register");
+
+    return has_feature;
+}
+
+/*
+ * Some features are identified by more than one nibble.
+ * In this case we need to add extra mappings between features and nibble values
+ */
+static uint32 feature_ids[] = {
+    (FEATURE_PAUTH << 16) | DEF_FEAT(AA64ISAR1, 1, 1, FEAT_GR_EQ), /* APA (QARMA5) */
```

Comment:
I don't understand: which feature is this?  FEATURE_PAUTH?  What is "APA (QARMA5)"?  Is it that FEATURE_PAUTH really means any one of 5 different sub-features?  Reword to `FEATURE_PAUTH.APA_QARMA5` or sthg to make it more clear?  Why aren't the FEATURE_PAUTH all grouped together in this array?

Reply:
I have expanded the comments and grouped the rows by feature to hopefully make this clearer.

The Arm ARM states "If the value of one or more of the ID_AA64ISAR1_EL1.{GPI, GPA, API, APA} or ID_AA64ISAR2_EL1.{GPA3, APA3} fields is nonzero, then pointer authentication is implemented."

So as well as the nibble in the definition of FEATURE_PAUTH, we have to add the 5 more in this table.